sdk/core/azure-core/src/main/java/com/azure/core/http/AuthorizationChallengeHandler.java

```diff
@@ -0,0 +1,402 @@
+            hexStringOf(digest.digest(requestEntity))).getBytes(StandardCharsets.UTF_8)));
+    }
+
+    /*
+     * Calculates the 'response' hex string when qop is unknown.
+     *
+     * This performs the following operations:
+     * - Create the digest of (ha1 + ":" + nonce + ":" + ha2).
+     * - Return the resulting bytes as a hex string.
+     */
+    private String calculateResponseUnknownQop(MessageDigest digest, String ha1, String nonce, String ha2) {
+        return hexStringOf(digest.digest(String.format("%s:%s:%s", ha1, nonce, ha2).getBytes(StandardCharsets.UTF_8)));
+    }
+
+    /*
+     * Calculates the 'response' hex string when 'qop=auth' or 'qop=auth-int'.
+     *
+     * This performs the following operations:
+     * - Create the digest of (ha1 + ":" + nonce + ":" + nc + ":" + cnonce + ":" + qop + ":" + ha2).
+     * - Return the resulting byes as a hex string.
+     *
+     * nc, nonce count, is represented in a hexadecimal format.
+     */
+    private String calculateResponseKnownQop(MessageDigest digest, String ha1, String nonce, int nc, String cnonce,
+        String qop, String ha2) {
+        return hexStringOf(digest.digest(String.format("%s:%s:%08X:%s:%s:%s", ha1, nonce, nc, cnonce, qop, ha2)
+            .getBytes(StandardCharsets.UTF_8)));
+    }
+
+    /*
+     * Calculates the hashed username value if the authenticate challenge has 'userhash=true'.
+     */
+    private String calculateUserhash(MessageDigest digest, String realm) {
+        return Base64.getEncoder()
+            .encodeToString(digest.digest(String.format("%s:%s", username, realm).getBytes(StandardCharsets.UTF_8)));
+    }
+
+    /*
+     * Attempts to retrieve the digest for the specified algorithm.
+     */
+    private static MessageDigest getAlgorithmDigest(String algorithm) {
+        if (algorithm.endsWith(_SESS)) {
+            algorithm = algorithm.substring(0, algorithm.length() - _SESS.length());
+        }
+
+        try {
+            // The SHA-512/256 algorithm is sent back as SHA-512-256, convert it to its common name.
+            return MessageDigest.getInstance(SHA_512_256.equals(algorithm) ? "SHA-512/256" : algorithm);
+        } catch (NoSuchAlgorithmException e) {
+            return null;
+        }
+    }
+
+    /*
+     * Splits the Authenticate challenges by the algorithm it uses.
+     */
+    private static Map<String, List<HttpHeaders>> partitionByChallengeType(List<HttpHeaders> challenges) {
+        return challenges.stream()
+            .collect(Collectors.groupingBy(headers -> {
+                String algorithmHeader = headers.getValue(ALGORITHM);
+
+                // RFC7616 specifies that is the "algorithm" header is null it defaults to MD5.
+                return (algorithmHeader == null) ? MD5 : algorithmHeader.toUpperCase(Locale.ROOT);
+            }));
+    }
+
+    /*
+     * Creates a client nonce.
+     */
+    private static String generateCnonce() {
+        return "0a4f113b";
```

Comment:
Does it have to be a new nonce for every new request?

Reply:
Yeah! This was hard coded to verify that a test case worked, I'll update this with something pluggable for testing.